fw/http_msg.c

```diff
@@ -1520,11 +1469,12 @@ __tfw_http_msg_expand_from_pool(TfwHttpResp *resp, const TfwStr *str,
 			if (unlikely(skb_room == 0 || nr_frags == MAX_SKB_FRAGS))
 			{
```

Comment:
The current implementation aimed to save much more memory as possible, but I'm not sure that it worth and that right algorithm was chosen(too many unnecessary work on each copy) and looks like we can sacrifice a little bit memory to achieve better performance.

Suggested approach has some disadvantages for me as well. Using `TfwStr` we anyway should to construct `TfwStr`, track the size of it, iterate over all chunks and call copy function for each chunk. This is also a lot of work that will introduce some overhead.

[SPDY](https://www.chromium.org/spdy/spdy-whitepaper/)'s research say:

> Request headers today vary in size from ~200 bytes to over 2KB

This means we can allocate from pool buffer with size 2KB or even 4KB only for headers, at this stage we don't place it as fragment to the SKB, then construct all headers inside this buffer, when all headers are constructed just add buffer to the SKB as paged fragment. With this approach we avoid all SKBs related checks on every copy doing all work in-place without any temporary buffers.

Reply:
I'm not sure if I understood you correctly: do you propose to expand HTTP messages with raw headers, without TfwStr? We have similar approach to write/adjust messages w/o forming TfwStr and/or filling HTTP headers table, but this works only if we do not modify a message later. I see that `*_msg_expand_from_pool*(` functions are called in many places, so it's possible (is it?) that we need to do some actions with the expanded messages later.

Reply:
I propose to use allocated block of memory instead of that:
> create a per-cpu static TfwStr of chunks N, where N is say the current number of chunks in total processed by tfw_h1_adjust_req() for any real life request

and just copy headers as we do now. This needs to avoid all unnecessary checks mentioned here:

> but in most cases all the checks for compound and duplicate strings, not enough space in the pool and inside the pool allocator as well as the skb checks are ran in vain

Reply:
This comment will be addressed in https://github.com/tempesta-tech/tempesta/issues/2136